node/impl/full/txhashmanager.go

```diff
@@ -23,6 +23,49 @@ func (m *EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) e
 	return nil
 }
 
+// FillIndexGap back-fills the gap between the current head and the last populated transaction index
+func (m *EthTxHashManager) FillIndexGap(ctx context.Context) error {
+	lastTxHash, lastTxCid, err := m.TransactionHashLookup.GetLastTransaction()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Start filling transaction index gap from: %s", lastTxHash)
+
+	totalProcessedBlock := 0
+	ts := m.StateAPI.Chain.GetHeaviestTipSet()
+	for _, block := range ts.Blocks() {
```

Comment:
I think we need to keep going back in the chain to backfill. This loop terminates after backfilling messages in the heaviest tipset only.

Reply:
We need to be careful on two fronts with going backward up the chain for two reasons:

1. the last processed message CID may not be on the canonical chain anymore so you may never encounter it; and
2. the last processed message may not even be in the blockstore due to a splitstore compaciton or some other _happening_ (e.g. you've copied across someone else's txhash.db or restored an old one from disk, whatever)

I suspect `ChainGetPath` might be a good API to try here: get the tipset of the `lastTxHash`, get the latest tipset, find the path between. I'm just not sure what the error conditions are for that - how big a gap is too large, and how do we deal with splitstore GC problems (I suspect "find me the tipset for `lastTxHash` would error in that case if it was too old, in which case what's the strategy then?).

Reply:
@rvagg I have been going through the `ChainGetPath` code, but I didn't see it returning a `too old` error, it only returns a `not found` error.

If it is not found then nodes have to import the snapshot.
Is there any other way to get the state?

Also `ChainGetPath` returns both `revert` and `apply` changes, so do we have to filter out `apply` changes and start backfilling?

Reply:
Discuss this with @akaladarshi  async. We worked through multiple solutions here. He will post a note about it.